README.md
# grunt-loop-mocha

A simple wrapper for running multiple mocha instances from a single grunt task target.

## Abilities

1. run mocha multiple times from a single target
2. set stringified JSON environment variables, customized for each mocha run

I wrote this module to facilitate selenium webdriver testing. For that, each mocha run is a different browser target
(specified in the iterations object of the config). My selenium test scripts
use stringified JSON environment variables to configure the test runs. Perhaps you'll find it useful for this or other purposes.

## Example usage

Below is the config block for this task from gruntfile.js

```javascript
/* ... */
loopmocha: {
	src: ["test/*.js"],
	options: {
		mocha: {
			parallel: true,
			globals: ['should'],
			timeout: 3000,
			ui: 'bdd',
			reporter: "xunit-file",
		},
		loop: {
		    reportLocation: "test/report"
		},
		env1: {
			stringVal: "fromfile"
		},
		env2: {
			jsonVal: {
				foo: {
					bar: {
						stringVal: "baz"
					}
				}
			}
		}
		iterations: [
			{
				"description": "first",
				"env1": {
					"someKey": "some value"
				}
			},
			{
				"description": "second",
				"env2": {
					"someOtherKey": "some other value"
				}
			},
			{
				"description": "third",
				"mocha": {
					"timeout": 4000
				}
			},
			{
				"description": "fifth",
				"env1": {
					"anotherKey": "BLERG"
				},
				"env2": {
					"yetAnotherKey": 123
				}

			}
		]
	}
}
/* ... */

grunt.registerTask('test', 'loopmocha');
```

### mocha options

any supported mocha command line argument is accepted.

### loop options

* parallel (optional: defaults to false): To engage a parallel testing ability, specify parallel.type = "file|directory". Optionally specify parallel.limit to limit the concurrent running processes
* reportLocation (required if using xunit-file reporter): specify where xunit report files should be written. Note: if you are using "xunit-file" as your reporter, you need to add it to your package.json
* noFail (optional: defaults to false): If true, the task will exit as zero regardless of any mocha test failures

### iterations options
Array of JSON objects. mocha will loop for each item, using its properties for the mocha run
  * iterations[N].description (optional): put this within your iteration objects in order to better label your console output or xunit report file
  * iterations[N].mocha (optional): grunt-loop-mocha will merge this mocha object with the one in your root "options" namespace for this iteration's mocha run
  * iterations[N].<mocha process env variables>: grunt-loop-mocha will merge this mocha object with any of the same name it finds in your root "options" namespace


### mocha process environment variables
Additional sibling elements to mocha/loop/iterations will be set as environment variables of the same name as the key. If the key value is a JSON object,
grunt-loop-mocha will call JSON.stringify before storing it as an

### mocha_bin option (optional: defaults to null)
String, path to mocha or 'mocha compatible' executable. This option allows to override default mocha
```javascript
     loopmocha.options."mocha_bin": '<custom-bin-path>'
```


CHANGELOG.md
# grunt-loop-mocha Changes

## v1.2.2

* documentation for `mocha_bin` feature. See: https://github.com/grawk/grunt-loop-mocha/pull/31
* update peer dependency on grunt to support grunt 1.0: https://github.com/grawk/grunt-loop-mocha/pull/33
* add contributors to package.json

## v1.2.1

* **Enhancement** remove manual directory creation for the xunit-file case since xunit-file now handles that internally

## v1.2.0

* **New feature** adding configurable option for path to mocha binary
* **New feature** better formatting for console output

## v1.1.0

**Bug fix** Fixes #24. Minor release bump since this bug fix does represent a change in behavior. Though presumably nobody was relying
on the buggy behavior. It just took some time to find this edge case due to the lack of a comprehensive unit test suite.

## v1.0.0

**Breaking change** moving all non-mocha options out of "mocha" config namespace (e.g. reportLocation). Adding "loop" namespace for such options.
See https://github.com/grawk/grunt-loop-mocha/blob/master/README.md#loop-options for more details

**New feature** parallelism by file or directory in addition to parallelism by iterations. See [parallel option](https://github.com/grawk/grunt-loop-mocha/blob/1.0.development/README.md#loop-options)

## v0.3.0

* Change structure of JSON config to namespace the mocha options, and namespace additional values as siblings to the mocha/iterations objects
* Pass-through variables no longer set via argv on the command line. Instead, per the additional namespaces on the options object, stringified JSON environment variables will be set

## v0.2.6

* Export iteration label as a command line arg for use in mocha tests

## v0.2.5

* Modify parallel flag handling such that options.parallel === undefined will not cause an exception

## v0.2.4

* Modify parallel flag handling such that Boolean true as well as any string version of "true" will evaluate to "true". Allows flag to be passed in via command line arg

## v0.2.3

* Modify noFail flag handling such that Boolean true as well as any string version of "true" will evaluate to "true". Allows flag to be passed in via command line arg

## v0.2.2

* Adding noFail option to prevent mocha from returning non-zero to the parent shell process. Allows a continuous integration build to determine success/failure based on xunit output

## v0.2.1

* Adding parallel mocha run option. Set "parallel": true in your Gruntfile config. Value will default to false. Results in use of "async.forEach" versus "async.forEachSeries" to execute mocha.

test/test.js
"use strict";
before(function(done) {
	//some setup
	done();
});
after(function(done) {
	//some teardown
	done();
});
describe("Mocha Tests", function() {
	it("should run", function(done) {
		done();
	})
})

test/test2.js
"use strict";
before(function(done) {
	//some setup
	done();
});
after(function(done) {
	//some teardown
	done();
});
describe("Mocha Tests 2", function() {
	it("should run", function(done) {
		done();
	})
})

test/process-loop-test.js
"use strict";
describe("Process Loop", function () {
  var process = require('../tasks/process-loop.js')
  var noop = function noop() {
  }
  var expect = require('chai').expect
  var EventEmitter = require('events').EventEmitter

  it("should call @spawnOnce@", function (done) {

    var testMe = process({log: {writeln: noop}}
      , function (path, op, env) {
        // test here
        expect(path)
          .to.equal('some/mocha/path')
        expect(op)
          .to.deep.equal(['other', 'options', 'one', 'two'])
        expect(env)
          .to.have.deep.property('env.opt')
          .that.equals('"process"')
        expect(env)
          .to.have.deep.property('env.XUNIT_FILE')
          .that.equals('./a/report/location/xunit-Test Label.xml')

        // We need to hand some stuff back...
        var ee = new EventEmitter()
        ee.stdout = {pipe: noop}
        ee.stderr = {pipe: noop}
        setTimeout(function () {
          ee.emit('close', 0)
        }, 1)
        return ee
      })

    testMe({filesSrc: ['one', 'two'], mocha_path: 'some/mocha/path', reportLocation: './a/report/location', localopts: ['other', 'options'], localOtherOptionsStringified: {opt: '"process"'}, itLabel: 'Test Label', localMochaOptions: {
        reporter: "xunit-file"
      }, loopOptions: {
        reportLocation: "/some/path"
      }}
      , function (err, results) {
        expect(results)
          .to.deep.equal([
            [ 0, 'Test Label' ]
          ])
        done()
      })
  })

  it("should call spawn @onceEachFile@", function (done) {
    var testMe = process({log: {writeln: noop}}
      , function (path, op, env) {
        // test here
        expect(path)
          .to.equal('some/mocha/path')
        expect(env)
          .to.have.deep.property('env.opt')
          .that.equals('"process"')

        if (op.indexOf('one') >= 0) {
          expect(op)
            .to.deep.equal(['other', 'options', 'one'])
          expect(env)
            .to.have.deep.property('env.XUNIT_FILE')
            .that.equals('./a/report/location/xunit-Test Label-one.xml')

        } else if (op.indexOf('two') >= 0) {
          expect(op)
            .to.deep.equal(['other', 'options', 'two'])
          expect(env)
            .to.have.deep.property('env.XUNIT_FILE')
            .that.equals('./a/report/location/xunit-Test Label-two.xml')
        } else {
          expect(true).to.not.be.true
        }

        // We need to hand some stuff back...
        var ee = new EventEmitter()
        ee.stdout = {pipe: noop}
        ee.stderr = {pipe: noop}
        setTimeout(function () {
          ee.emit('close', 0)
        }, 1)
        return ee
      })

    testMe({filesSrc: ['one', 'two'], mocha_path: 'some/mocha/path', reportLocation: './a/report/location', localopts: ['other', 'options'], localOtherOptionsStringified: {opt: '"process"'}, itLabel: 'Test Label', localMochaOptions: {
        reporter: "xunit-file"
      }, loopOptions: {
        parallel: {type: "file"}, reportLocation: "/some/path"
      }}
      , function (err, results) {
        expect(results)
          .to.deep.equal([
            [ 0, 'Test Label-one' ],
            [ 0, 'Test Label-two' ]
          ])
        done()
      })
  })

  it("should call spawn @onceEachDirectory@", function (done) {
    var testMe = process({log: {writeln: noop}}
      , function (path, op, env) {
        // test here
        expect(path)
          .to.equal('some/mocha/path')
        expect(env)
          .to.have.deep.property('env.opt')
          .that.equals('"process"')

        if (op.indexOf('/dir1/one') >= 0) {
          expect(op)
            .to.deep.equal(['other', 'options', '/dir1/one', '/dir1/two'])
          expect(env)
            .to.have.deep.property('env.XUNIT_FILE')
            .that.equals('./a/report/location/xunit-Test Label--dir1.xml')

        } else if (op.indexOf('/dir2/one') >= 0) {
          expect(op)
            .to.deep.equal(['other', 'options', '/dir2/one', '/dir2/two'])
          expect(env)
            .to.have.deep.property('env.XUNIT_FILE')
            .that.equals('./a/report/location/xunit-Test Label--dir2.xml')
        } else {
          expect(true).to.not.be.true
        }

        // We need to hand some stuff back...
        var ee = new EventEmitter()
        ee.stdout = {pipe: noop}
        ee.stderr = {pipe: noop}
        setTimeout(function () {
          ee.emit('close', 0)
        }, 1)
        return ee
      })

    testMe({filesSrc: ['/dir1/one', '/dir1/two'
        , '/dir2/one', '/dir2/two'], mocha_path: 'some/mocha/path', reportLocation: './a/report/location', localopts: ['other', 'options'], localOtherOptionsStringified: {opt: '"process"'}, itLabel: 'Test Label', localMochaOptions: {
        reporter: "xunit-file"
      }, loopOptions: {
        parallel: {type: "directory"}, reportLocation: "/some/path"
      }}
      , function (err, results) {
        expect(results)
          .to.deep.equal([
            [ 0, 'Test Label--dir1' ],
            [ 0, 'Test Label--dir2' ]
          ])
        done()
      })
  })
})

tasks/loop-mocha.js
/*───────────────────────────────────────────────────────────────────────────*\
 │  Copyright (C) 2016 PayPal                                                  │
 │                                                                             │
 │                                                                             │
 │   Licensed under the Apache License, Version 2.0 (the "License"); you may   │
 │   not use this file except in compliance with the License. You may obtain   │
 │   a copy of the License at http://www.apache.org/licenses/LICENSE-2.0       │
 │                                                                             │
 │   Unless required by applicable law or agreed to in writing, software       │
 │   distributed under the License is distributed on an "AS IS" BASIS,         │
 │   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  │
 │   See the License for the specific language governing permissions and       │
 │   limitations under the License.                                            │
 \*───────────────────────────────────────────────────────────────────────────*/
/*
 * grunt-loop-mocha
 * https://github.com/grawk/grunt-loop-mocha
 *
 * Copyright (c) 2013 Matt Edelman
 * Licensed under the MIT license.
 */
"use strict";

module.exports = function (grunt) {

  var path = require('path'),
    fs = require('fs'),
    util = grunt.util,
    nutil = require('util'),
    child_process = require("child_process"),
    _ = require('lodash'),
    async = require('async'),
    exists = grunt.file.exists,
    iterationError = false,
    iterationRemaining,
    noFail = false,
    iterationResults = {};

  grunt.registerMultiTask('loopmocha', 'Run mocha multiple times', function () {
    var options = this.options(),
      mochaDefaultOptions = grunt.config.get("loopmocha.options.mocha"),
      mochaOptions = _.merge(mochaDefaultOptions, options.mocha),
      loopDefaultOptions = grunt.config.get("loopmocha.options.loop") || {},
      loopOptions = _.merge(loopDefaultOptions, options.loop),
      allDefaultOptions = grunt.config.get("loopmocha.options"),
      otherDefaultOptions = {},
      otherOptions = {},
      reportLocation = loopOptions.reportLocation || '',
      asyncMethod = (loopOptions.parallel && loopOptions.parallel !== false) ? "map" : "mapSeries",
      binPath = (options.mocha_bin || '.bin/mocha') + (process.platform === 'win32' ? '.cmd' : ''),
      mocha_path = path.join(__dirname, '..', '/node_modules/', binPath),
      iterations = options.iterations || undefined,
      iterationRemaining = Object.keys(iterations).length,
      iterationIndex = 0,
      done = this.async(),
      filesSrc = this.filesSrc,
      runStamp = (new Date()).getTime();

    //catalog any extra default options
    _.each(_.omit(allDefaultOptions, 'src', 'basedir', 'mocha', 'iterations'), function (value, key) {
      otherDefaultOptions[key] = value;
    });
    //catalog any extra task specific options
    _.each(_.omit(options, 'src', 'basedir', 'mocha', 'iterations'), function (value, key) {
      otherOptions[key] = value;
    });
    //merge together the default and task specific extra options
    _.merge(otherOptions, otherDefaultOptions);
    if (!exists(mocha_path)) {
      var i = module.paths.length,
        bin;
      while (i--) {
        bin = path.join(module.paths[i], binPath);
        if (exists(bin)) {
          mocha_path = bin;
          break;
        }
      }
    }

    if (!exists(mocha_path)) {
      grunt.fail.warn('Unable to find mocha.');
    }


    // map or mapLimit.  The cb will take an err and result...
    async[asyncMethod](iterations, function (el, cb) {
      var i,
        opts = {},
        localopts = [],
        localOtherOptions = {},
        mergedOtherOptions = {},
        localOtherOptionsStringified = {},
        localMochaOptions = (el.mocha) ? _.merge(mochaOptions, el.mocha) : mochaOptions,
        itLabel = runStamp + "-" + ((el.description) ? (el.description) : (++iterationIndex)); // = opts_array.slice(0);

      //merge the iteration extra options with the default/task specific extra options
      _.each(_.omit(el, 'description', 'mocha'), function (value, key) {
          console.log('adding key', key, 'value', value);
        localOtherOptions[key] = value;
      });
        _.merge(mergedOtherOptions, otherOptions, localOtherOptions);

        //stringify the extra options for passage via env
      _.each(mergedOtherOptions, function (value, key) {
        console.log("[grunt-loop-mocha] setting ENV var ", key, "with value", value);
        localOtherOptionsStringified[key] = (value.constructor === Object) ? JSON.stringify(value) : value;
      });

      // put the localMochaOptions into opts so we can pass
      // to mocha with --thing.  Remove options grunt-loop-mocha
      // is supporting
      _.each(_.omit(localMochaOptions
        , 'reportLocation'
        , 'iterations'
        , 'limit' // the limit var for mapLimit
      )
        , function (value, key) {
          if (value !== 0) {
            opts[key] = value || "";
          }
        });
      if (loopOptions.noFail && loopOptions.noFail.toString().toLowerCase() === "true") {
        noFail = true;
      }

      Object.keys(opts).forEach(function (key) {
        if (key === "description") {
          return;
        }
        localopts.push("--" + key);
        if (opts[key] !== "" && opts[key] !== undefined) {
          localopts.push(opts[key]);
        }
      });

      // spill off the processes.  This can be quite a few.
      // the results will be in the form
      // [[returnCode, itterationName], ...]
      require('./process-loop.js')(grunt)({
          filesSrc: filesSrc, mocha_path: mocha_path, reportLocation: reportLocation, localopts: localopts, localOtherOptionsStringified: localOtherOptionsStringified, itLabel: itLabel, localMochaOptions: localMochaOptions, loopOptions: loopOptions
        }
        , cb);

    }, function (err, results) {
      var iterationResults = results.reduce(function (results, runs) {
        runs.forEach(function (run) {
          results[run[1]] = run[0];
          iterationError = iterationError || !!run[0];
        });
        return results;
      }, {});
      if (iterationError) {
        var msg = "[grunt-loop-mocha] error, please check erroneous iteration(s): \n" + JSON.stringify(iterationResults, null, 4) + "\n";
        if (noFail === true) {
          console.log(msg);
          done();
        }
        else {
          done(new Error(msg));
        }

      } else {
        done();
      }
      console.log("[grunt-loop-mocha] Total Runtime", Math.floor((((new Date()).getTime()) - runStamp)/1000) + "s");
    });
  });
};


tasks/process-loop.js
/*───────────────────────────────────────────────────────────────────────────*\
 │  Copyright (C) 2016 PayPal                                                  │
 │                                                                             │
 │                                                                             │
 │   Licensed under the Apache License, Version 2.0 (the "License"); you may   │
 │   not use this file except in compliance with the License. You may obtain   │
 │   a copy of the License at http://www.apache.org/licenses/LICENSE-2.0       │
 │                                                                             │
 │   Unless required by applicable law or agreed to in writing, software       │
 │   distributed under the License is distributed on an "AS IS" BASIS,         │
 │   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  │
 │   See the License for the specific language governing permissions and       │
 │   limitations under the License.                                            │
 \*───────────────────────────────────────────────────────────────────────────*/

'use strict';

var spawn = require("child_process").spawn;
var path  = require('path');
var util  = require('util');
var _     = require('lodash');
var async = require('async');

/**
 * Main level.  Grunt is injected
 * from the task.  _spawn is injected
 * to help with tests.
 */
module.exports = function exports (grunt, _spawn) {

  // injection here
  if (_spawn) {
    spawn = _spawn;
  }
  // Give us a function that will work with async
  return function processLoop(op, cb) {
    function work(_itLabel, _filesSrc, _env, _op, _cb) {

      // inform the world that we are going to start
      grunt.log.writeln("[grunt-loop-mocha] iteration: ", _itLabel);

      // update the reporter file
      if (localMochaOptions.reporter === "xunit-file") {
        // Only update the env for the spawned process.
        _env.XUNIT_FILE = reportLocation + "/xunit-" + _itLabel + ".xml";
        grunt.log.writeln("[grunt-loop-mocha] xunit output: ", _env.XUNIT_FILE);
      }

      // push the files into localopts
      _filesSrc.forEach(function (el) {
        _op.push(el);
      });

      // more notify
      grunt.log.writeln("[grunt-loop-mocha] mocha argv: ", _op.toString());

      // start a process
      var child = spawn(mocha_path
        , _op
        , {env: _env});

      // pipe the output (in paralell this is going to be noisey)
      child.stdout.pipe(process.stdout);
      child.stderr.pipe(process.stderr);

      // report back the outcome
      child.on('close', function (code) {
        _cb(null, [code, _itLabel]);
      });
    }
    // rip up the op into vars
    var filesSrc                      = op.filesSrc
      , mocha_path                    = op.mocha_path
      , reportLocation                = op.reportLocation
      , localopts                     = op.localopts
      , localOtherOptionsStringified  = op.localOtherOptionsStringified
      , itLabel                       = op.itLabel
      , localMochaOptions             = op.localMochaOptions
      , loopOptions                   = op.loopOptions;

    console.log('loopOptions', loopOptions);
    var limit         = (loopOptions.parallel && loopOptions.parallel.limit) ? loopOptions.parallel.limit : 5;
    var parallelType  = (loopOptions.parallel && loopOptions.parallel.type) ? loopOptions.parallel.type : "none";
    var env           = _.merge(process.env, localOtherOptionsStringified);

    // pick a way to split up the work
    if (parallelType === 'directory') {   // async by directory
      async
        .mapLimit(_.chain(filesSrc)
        // group by the directory path
        .groupBy(function(file) {
          return path.dirname(file);
        })
        // rework the data into something async will be ok with
        .map(function(files, dir) {
          return [files, dir];
        })
        .value()
        , limit
        , function(args, _cb) {
          // update the label, do the work
          work(itLabel + '-' + args[1].replace(/[\/|\\|:]/g, '-')
            , args[0]
            , _.cloneDeep(env)
            , _.cloneDeep(localopts)
            , _cb);
        }
        , cb);
    } else if (parallelType === 'file') { // async by file
      async
        .mapLimit(filesSrc
        , limit
        , function(file, _cb) {
          // update the label, do the work
          work(itLabel + '-' + file.replace(/[\/|\\|:]/g, '-')
            , [file]
            , _.cloneDeep(env)
            , _.cloneDeep(localopts)
            , _cb);
        }
        , cb);
    }  else {                             // not async
      work(itLabel, filesSrc, env, localopts
        , function(err, result) {
          // to normalize results, we wrap this one in an
          // array, even though there is only one.
          cb(err, [result]);
        });
    }
  };
};

Gruntfile.js
module.exports = function (grunt) {

	// Project configuration.
	grunt.initConfig({
		watch: {
			files: '<config:lint.files>',
			tasks: 'default'
		},
		//Mocha configuration
		loopmocha: {
			src: ["test/*.js"],
			options: {
				iterations: [
					{
						"description": "first",
						"env1": {
							"someKey": "some value"
						}
					},
					{
						"description": "second",
						"env2": {
							"someOtherKey": "some other value"
						}
					},
					{
						"description": "third",
						"mocha": {
							"timeout": 4000
						}
					},
					{
						"description": "fifth",
						"env1": {
							"anotherKey": "BLERG"
						},
						"env2": {
							"yetAnotherKey": 123
						}

					},
					{
						"description": "many",
						"env1": {
							"anotherKey": "BLERG"
						},
						"env2": {
							"yetAnotherKey": 123
						}

					}
				],
				mocha: {
					parallel: "file",
					globals: ['should'],
					timeout: 3000,
					ui: 'bdd',
					reporter: "xunit-file"
        },
        loop: {
          reportLocation: "test/report",
          parallel: "iteration"
        },
				env1: {
					stringVal: "fromfile"
				},
				env2: {
					jsonVal: {
						foo: {
							bar: {
								stringVal: "baz"
							}
						}
					}
				}


			}
		},
		jshint: {
			options: {
				curly: true,
				eqeqeq: true,
				immed: true,
				latedef: true,
				newcap: true,
				noarg: true,
				sub: true,
				undef: true,
				boss: true,
				eqnull: true,
        laxcomma: true,
				node: true,
				es5: true,
				globals: {}
			},

			files: ['grunt.js', 'tasks/**/*.js']
		}
	});

	// Load local tasks.
	grunt.loadTasks('tasks');
	grunt.loadNpmTasks('grunt-contrib-jshint');

	// Default task.
	grunt.registerTask('test', 'loopmocha');
	grunt.registerTask('default', ['jshint', 'test']);
};